File: frontend/src/types/procurement.ts
// ── Procurement types ─────────────────────────────────────────────

export type Priority = "LOW" | "NORMAL" | "HIGH" | "URGENT";

export type PurchaseRequestStatus = "DRAFT" | "VALIDATED" | "APPROVED" | "REJECTED";

export interface ExtractedRequisition {
  item: string;
  quantity: number;
  delivery_location: string;
  required_date: string; // YYYY-MM-DD
  priority: Priority;
}

export interface ExtractionResultResponse {
  raw_message: string;
  extracted: ExtractedRequisition | null;
  is_valid: boolean;
  validation_errors: Record<string, string> | null;
}

export interface CreatePurchaseRequestRequest {
  item: string;
  quantity: number;
  delivery_location: string;
  required_date: string; // YYYY-MM-DD
  priority: Priority;
  raw_message?: string | null;
}

export interface ProductResponse {
  id: string;
  sku: string;
  name: string;
  unit: string;
}

export interface PurchaseRequestItem {
  id: string;
  product: ProductResponse;
  quantity: number;
}

export interface PurchaseRequest {
  id: string;
  request_code: string;
  requested_by_user_id: string;
  delivery_location: string;
  required_date: string; // ISO 8601
  priority: Priority;
  status: PurchaseRequestStatus;
  raw_chat_input?: string | null;
  extracted_json?: Record<string, unknown> | null;
  items: PurchaseRequestItem[];
  created_at: string; // ISO 8601
  updated_at: string; // ISO 8601
}

export interface SupplierResponse {
  id: string;
  supplier_code: string;
  name: string;
  city: string | null;
  quality_score: number;
  delivery_score: number;
  capacity_score: number;
  lead_time_days: number;
  is_active: boolean;
}

export interface SupplierScoreBreakdown {
  cost_score: number;
  quality_score: number;
  delivery_score: number;
  capacity_score: number;
  lead_time_score: number;
  overall_score: number;
  reasons: string[];
}

export interface SupplierRecommendation {
  supplier: SupplierResponse;
  product_id: string;
  unit_price: number;
  available_capacity: number;
  lead_time_days: number;
  score_breakdown: SupplierScoreBreakdown;
  is_recommended: boolean;
}

export interface SupplierRecommendationsResponse {
  purchase_request_id: string;
  request_code: string;
  recommendations: SupplierRecommendation[];
}

export interface PurchaseOrderItem {
  id: string;
  product_id: string;
  quantity: number;
  unit_price: number;
  line_total: number;
  product?: ProductResponse;
}

export interface POShipmentSummary {
  id: string;
  shipment_code: string;
  tracking_number: string;
  status: string;
}

export interface POTruckSummary {
  id: string;
  truck_code: string;
  trailer_id: string;
  status: string;
  current_eta: string | null;
}

export interface PurchaseOrderResponse {
  id: string;
  po_code: string;
  purchase_request_id: string;
  total_amount: number;
  delivery_location: string;
  expected_delivery_date: string;
  status: string;
  recommendation_score: number | null;
  created_at: string;
  
  supplier: SupplierResponse;
  items: PurchaseOrderItem[];
  shipment: POShipmentSummary | null;
  truck: POTruckSummary | null;
}

File: backend/app/schemas/procurement.py
from datetime import date, datetime
from enum import Enum
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field, constr, conint


class Priority(str, Enum):
    LOW = "LOW"
    NORMAL = "NORMAL"
    HIGH = "HIGH"
    URGENT = "URGENT"


class PurchaseRequestStatus(str, Enum):
    DRAFT = "DRAFT"
    VALIDATED = "VALIDATED"
    APPROVED = "APPROVED"
    REJECTED = "REJECTED"


class RequisitionChatRequest(BaseModel):
    message: str = Field(..., min_length=5, max_length=1000)


class ExtractedRequisition(BaseModel):
    item: str = Field(..., min_length=1)
    item_description: Optional[str] = None
    quantity: int = Field(..., gt=0, le=100000)
    delivery_location: str = Field(..., min_length=1)
    required_date: date
    priority: Priority = Priority.NORMAL

    def model_post_init(self, __context: Any) -> None:
        if not self.item_description and self.item:
            self.item_description = self.item
        elif self.item_description and not self.item:
            self.item = self.item_description


class ExtractionResultResponse(BaseModel):
    raw_message: str
    extracted: Optional[ExtractedRequisition] = None
    is_valid: bool = False
    validation_errors: Optional[Dict[str, Any]] = None
    confidence: Optional[Dict[str, float]] = None
    missing_fields: Optional[List[str]] = Field(default_factory=list)
    warnings: Optional[List[str]] = Field(default_factory=list)


class CreatePurchaseRequestRequest(BaseModel):
    item: Optional[str] = None
    item_description: Optional[str] = None
    quantity: int = Field(..., gt=0, le=100000)
    delivery_location: str = Field(..., min_length=1)
    required_date: date
    priority: Priority = Priority.NORMAL
    raw_message: Optional[str] = None

    def model_post_init(self, __context: Any) -> None:
        if not self.item and self.item_description:
            self.item = self.item_description
        elif not self.item_description and self.item:
            self.item_description = self.item


class ProductResponse(BaseModel):
    id: str
    sku: str
    name: str
    unit: str

    model_config = {"from_attributes": True}


class PurchaseRequestItemResponse(BaseModel):
    id: str
    product: ProductResponse
    quantity: int

    model_config = {"from_attributes": True}


class PurchaseRequestResponse(BaseModel):
    id: str
    request_code: str
    requested_by_user_id: str
    delivery_location: str
    required_date: datetime
    priority: str
    status: str
    raw_chat_input: Optional[str] = None
    extracted_json: Optional[Dict[str, Any]] = None
    items: List[PurchaseRequestItemResponse]
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}
